File: SysOp/Java/GM.java
package SysOp.Java;

public class GM {
    private int tamPag;
    private boolean[] pags;

    public GM(int tamM, int tamP){
        tamPag = tamP;
        pags = new boolean[tamM/tamPag];
        for(int i = 0; i < pags.length; i++) pags[i] = false;}

    boolean aloca(int numPalvaras, int[] tabPag){
        int tam = (numPalvaras/tamPag) + (((numPalvaras % tamPag) > 0) ? 1 : 0); //quantas paginas o programa precisa
        int count = tam;

        for(int i = 0; (i < pags.length) && (count > 0); i++) //verfica se tem todas as paginas necessarias
            if(!pags[i]) 
                count--;

        if(count != 0) return false;

        tabPag = new int[tam];
        for(int i = 0 ; count < tabPag.length; i++) //coloca as paginas alocadas no tabPag;
            if(!pags[i]){
                pags[i] = true;
                tabPag[count++] = i;}
        return true;}

    void desaloca(int[] tabPag){
        for(int i = 0; i < tabPag.length; i++)
            pags[tabPag[i]] = false;}
}

File: SysOp/cpp/gm.cpp
#include "gm.hpp"


GM::GM(int tamM, int tamP){
    tamPag = tamP;
    pags.resize(tamM/ tamPag, false);
}

bool GM::aloca(int numPalavras, vector<int>& tabPag){
    int tam = (numPalavras / tamPag) + ((numPalavras % tamPag) > 0 ? 1 : 0); //quantas paginas o programa precisa
    int count = tam;

    for(size_t i = 0; i < pags.size() && count > 0; i++){ //verfica se tem todas as paginas necessarias
        if(!pags[i])
            count--;
    }

    if(count != 0){ return false; }

    tabPag.clear();
    count = 0;

    for(size_t i = 0; count < tam && i < pags.size(); i++){
        if(!pags[i]){
            pags[i] = true;
            tabPag.push_back(i);
            count++;
        }
    }

    return true;

}

void GM::desaloca(vector<int>& tabPag){
    for(size_t i = 0; i < tabPag.size(); i++){
        if(tabPag[i] >= 0 && tabPag[i] < (int)pags.size()){
            pags[tabPag[i]] = false;
        }
    }
}

